p2p/unicast/router.go

```diff
@@ -0,0 +1,136 @@
+// TECHDEBT(#629): parameterize and expose via config.
+// readStreamTimeout is the duration to wait for a read operation on a
+// stream to complete, after which the stream is closed ("timed out").
+const readStreamTimeout = time.Second * 10
+
+var _ unicastRouterFactory = &UnicastRouter{}
+
+type unicastRouterFactory = modules.FactoryWithConfig[*UnicastRouter, *config.UnicastRouterConfig]
+
+type UnicastRouter struct {
+	base_modules.IntegratableModule
+
+	logger *modules.Logger
+	host   libp2pHost.Host
+	// messageHandler is the function to call when a message is received.
+	// host represents a libp2p network node, it encapsulates a libp2p peerstore
+	// & connection manager. `libp2p.New` configures and starts listening
+	// according to options.
+	// (see: https://pkg.go.dev/github.com/libp2p/go-libp2p#section-readme)
+	messageHandler typesP2P.MessageHandler
+	peerHandler    func(peer typesP2P.Peer) error
+}
+
+func Create(bus modules.Bus, cfg *config.UnicastRouterConfig) (*UnicastRouter, error) {
+	return new(UnicastRouter).Create(bus, cfg)
+}
+
+func (*UnicastRouter) Create(bus modules.Bus, cfg *config.UnicastRouterConfig) (*UnicastRouter, error) {
+	if err := cfg.IsValid(); err != nil {
+		return nil, err
+	}
+
+	rtr := &UnicastRouter{
+		logger:         cfg.Logger,
+		host:           cfg.Host,
+		messageHandler: cfg.MessageHandler,
+		peerHandler:    cfg.PeerHandler,
+	}
+	rtr.SetBus(bus)
+
+	// Don't handle incoming streams in client debug mode.
+	if !rtr.isClientDebugMode() {
+		rtr.host.SetStreamHandler(cfg.ProtocolID, rtr.handleStream)
+	}
+
+	return rtr, nil
+}
+
+// handleStream ensures the peerstore contains the remote peer and then reads
+// the incoming stream in a new go routine.
+func (rtr *UnicastRouter) handleStream(stream libp2pNetwork.Stream) {
+	rtr.logger.Debug().Msg("handling incoming stream")
+	peer, err := utils.PeerFromLibp2pStream(stream)
+	if err != nil {
+		rtr.logger.Error().Err(err).
+			Str("address", peer.GetAddress().String()).
+			Msg("parsing remote peer identity")
+
+		if err = stream.Reset(); err != nil {
+			rtr.logger.Error().Err(err).Msg("resetting stream")
+		}
+		return
+	}
+
+	if err := rtr.peerHandler(peer); err != nil {
+		rtr.logger.Error().Err(err).
+			Str("address", peer.GetAddress().String()).
+			Msg("adding remote peer to router")
+	}
+
+	go rtr.readStream(stream)
+}
+
+// readStream reads the message bytes out of the incoming stream and passes it to
+// the configured `rtr.messageHandler`. Intended to be called in a go routine.
+func (rtr *UnicastRouter) readStream(stream libp2pNetwork.Stream) {
+	// Time out if no data is sent to free resources.
+	// NB: tests using libp2p's `mocknet` rely on this not returning an error.
+	if err := stream.SetReadDeadline(newReadStreamDeadline()); err != nil {
+		// `SetReadDeadline` not supported by `mocknet` streams.
+		rtr.logger.Error().Err(err).Msg("setting stream read deadline")
```

Comment:
Should we return after logging the error?

Reply:
That would break tests which use the libp2p mocknet as per the comments. I think the only reason an error would happen is if a node was using an incompatible transport (like in the case of mocknet). Since each node explicitly supports specific transports via the multiaddr(s) it advertises, it is in control of which transports are used in this context. 

TL;DR this should only return an error if a node listens via an unsupported transport protocol (and in tests).

Reply:
Thoughts on leaving a comment: `// not returning an error for testing purposes. This should only produce an error if a node listens via an unsupported transport protocol, which should never happen in prod.`